## DMA 2023 ##
Make sure you fill in any place that says `YOUR CODE HERE` or `YOUR ANSWER HERE`, as well as your name below:

In [91]:
NAME = "Wajahat Khan"

# Lab 4: Neural Networks #
**Please read the following instructions very carefully**

## Working on the assignment / FAQs
- **Always use the seed/random_state as *42* wherever applicable** (This is to ensure repeatability in answers, across questions, students and coding environments).
- All questions will be graded manually.
- Most questions have two cells:
  - A code cell for your work/code
  - A text cell for giving your final answer
- The points each question carries are indicated.
- Most assignments have bonus questions for extra credit, do try them out!
- **Submitting the assignment** : Download the '.ipynb' and '.pdf' files from Colab and upload them to Gradescope. Do not delete any outputs from cells before submitting. Make sure to assign pages to questions when uploading your PDF to Gradescope.
- That's about it. Happy coding!


## About the dataset
This assignment uses a dataset obtained from the JSE Data Archive that contains biological and self-reported activity traits of a sample of college students at a single university uploaded in 2013. Background Information on the dataset: http://jse.amstat.org/v21n2/froelich/eyecolorgender.txt

For this lab, the dataset has already been split into a training set `df_train` and a test set `df_test`.


In [92]:
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC

from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.feature_extraction import DictVectorizer

from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, ParameterGrid

import numpy as np

import warnings
warnings.filterwarnings("ignore")

In [93]:
!wget http://askoski.berkeley.edu/~zp/lab_4_training.csv
!wget http://askoski.berkeley.edu/~zp/lab_4_test.csv

df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')

'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.


***
### Question 1 (1 point)
Calculate a baseline accuracy measure using the majority class, assuming a target variable of `gender`. The majority class is the most common value of the target variable in a particular dataset. Accuracy is calculated as (true positives + true negatives) / (all negatives and positives).

**Question 1.a**  
Find the majority class in the training set. If you always predicted this class in the training set, what would your accuracy be?

In [94]:
# YOUR CODE HERE
which_gender = df_train["gender"].value_counts().idxmax()
#print(df_train["gender"].value_counts())
Accuracy_caculation = (df_train['gender'] == which_gender).mean()

#This class method"
gender_counts =  df_train['gender'].value_counts()
accuracy_2 = gender_counts["female"] / len(df_train)
print(which_gender + " majority training set")
print(Accuracy_caculation)
print(accuracy_2)

female majority training set
0.5427852348993288
0.5427852348993288


**Answer: 0.5427852348993288**

**Question 1.b**   
If you always predicted this same class (majority from the training set) in the test set, what would your accuracy be?

In [95]:
# YOUR CODE HERE
which_gender = df_test["gender"].value_counts().idxmax()
Accuracy_caculation = (df_test['gender'] == which_gender).mean()
print(which_gender + " majority training set")
print(Accuracy_caculation)

female majority training set
0.5226130653266332


**Answer: 0.5226130653266332**

***
### Question 2 (1.5 points)###
Get started with Neural Networks.

   
Choose a NN implementation (we recommend Sklearn MLPclassifier) and specify which you choose. Be sure the implementation allows you to modify the number of hidden layers and hidden nodes per layer.  

NOTE: When possible, specify the logsig (`sigmoid`/`logistic`) function as the transfer function (another word for activation function) and use Levenberg-Marquardt backpropagation (`lbfgs`). It is possible to specify logistic in Sklearn MLPclassifier.  

**My NN implementation of choice: YOUR ANSWER HERE**

**Question 2.a**   
Train a neural network with a single 10 node hidden layer. Only use the `height` feature of the dataset to predict the `gender`. You will have to change `gender` to a 0 and 1 class. After training, use your trained model to predict the class (`gender`) using the `height` feature from the training set. What is the accuracy of this prediction?

In [96]:
## I am going to impor these files each time cause this dumb computer keeps turning gender into Nan values somehow
## I checked my code like 10 times and still can't tell when I rerun the cell it does that.
df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')

# YOUR CODE HERE
## I assume we can assign either or for the male and female
#df_train["gender"] = df_train["gender"].map({"male": 0, "female": 1})

df_train_copy = df_train.copy()
# Encode gender as 0 and 1 https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html
# pretty cool, I don't have keep doing the map stuff, pretty cool

le = LabelEncoder() 
df_train_copy['gender'] = le.fit_transform(df_train_copy['gender'])
X_train = df_train_copy[['height']]
y_train = df_train_copy['gender']


clf = MLPClassifier(hidden_layer_sizes=(10,), activation='logistic', solver='lbfgs', random_state=42)
clf.fit(X_train, y_train)
y_train_pred = clf.predict(X_train)
accuracy_train = accuracy_score(y_train, y_train_pred)
accuracy_train


0.8439597315436241

**Answer: 0.8439597315436241**

**Question 2.b (0.5 points)**  
Take the trained model from question 2.a and use it to predict the test set. This can be accomplished by taking the trained model and giving it the `height` feature values from the test set. What is the accuracy of this model on the test set?

In [97]:
# YOUR CODE HERE
## I am going to impor these files each time cause this dumb computer keeps turning gender into Nan values somehow
## I checked my code like 10 times and still can't tell when I rerun the cell it does that.

df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')

# YOUR CODE HERE
## I assume we can assign either or for the male and female
#df_test["gender"] = df_test["gender"].map({"male": 0, "female": 1})

# Create a copy of the test dataframe
df_test_copy = df_test.copy()
df_test_copy['gender'] = le.transform(df_test_copy['gender'])

X_test = df_test_copy[['height']]
y_test = df_test_copy['gender']

y_test_pred = clf.predict(X_test)
accuracy_test = accuracy_score(y_test, y_test_pred)
accuracy_test

0.8542713567839196

**Answer: 0.8542713567839196 **

**Question 2.c**   
Neural Networks tend to prefer smaller, normalized feature values. Try taking the log of the `height` feature in both training and testing sets or use a Standard Scalar operation in SKlearn to centre and normalize the data between 0-1 for continuous values. Repeat question 2.a and 2.b with the log version or the normalized and centered version of this feature.

In [61]:
# YOUR CODE HERE

## just having the copies instead of reimporting them might be better.
df_train_normalized = df_train.copy()
df_test_normalized = df_test.copy()

# scaler, StandardScaler thingy mucghtyyy
scaler = StandardScaler()
df_train_normalized['height'] = scaler.fit_transform(df_train_normalized[['height']])
df_test_normalized['height']  = scaler.transform(df_test_normalized[['height']])

# using the same LabelEncoder as before, 0 or 1, it assigns by itself instead of me mapping
df_train_normalized['gender'] = le.transform(df_train_normalized['gender'])
df_test_normalized['gender'] = le.transform(df_test_normalized['gender'])

# trainNormilized, height and Gender bruv --> Train
X_train_normalized = df_train_normalized[['height']]
y_train_normalized = df_train_normalized['gender']

# testNormilized, height and Gender ---> Test
X_test_normalized = df_test_normalized[['height']]
y_test_normalized = df_test_normalized['gender']

# Create and train the MLPClassifier with normalized height feature
clf = MLPClassifier(hidden_layer_sizes=(10,),
                               activation='logistic', max_iter = 200, verbose = 10, solver='lbfgs', random_state=42)
clf.fit(X_train_normalized, y_train_normalized)

## train, accruracyy
y_train_pred_normalized   = clf.predict(X_train_normalized)
accuracy_train_normalized = accuracy_score(y_train_normalized, y_train_pred_normalized)

## test. accruacy
y_test_pred_normalized   = clf.predict(X_test_normalized)
accuracy_test_normalized = accuracy_score(y_test_normalized, y_test_pred_normalized)

print(accuracy_train_normalized)
print(accuracy_test_normalized)


0.8439597315436241
0.8542713567839196


**Answer (accuracy on training set): 0.8439597315436241**
## Note I get a different answer with the same code on collab:  0.834731543624161

**Answer (accuracy on test set): 0.8542713567839196**
## Note I get a different answer with the same code on collab:  0.831658291457286

In [62]:
## I would like to go with 0.8347 as it seems more correct if you take a log, the nubmers should of changed? Idk whats going on

***

### Question 3 (1 point) ###
Many of the remaining features in the dataset are categorical. No ML method accepts categorical features, so transform `year`, `eyecolor`, `exercise` into a set of binary features, one feature per unique original feature value, and mark the binary feature as ‘1’ if the feature value matches the original value and ‘0’ otherwise. Using only these one-hot transformed features, train and predict the class of the test set. What was your accuracy using a Neural Network with a single 10 node hidden layer?

In [67]:
# YOUR CODE HERE

train_data = new_dftrain[['year', 'eyecolor', 'exercise']]
test_data = new_dftest[['year', 'eyecolor', 'exercise']]
encoded_train_data = pd.get_dummies(train_data, columns=["year", "eyecolor", "exercise"])
encoded_test_data = pd.get_dummies(test_data, columns=["year", "eyecolor", "exercise"])

X_train = encoded_train_data
Y_train = new_dftrain['gender']
classifier = MLPClassifier(hidden_layer_sizes=(10,), activation='logistic',
                    max_iter=200, solver='lbfgs', verbose=10, random_state=42)

classifier.fit(X_train, Y_train)
X_test = encoded_test_data
Y_test = new_dftest['gender']
y_pred = classifier.predict(X_test)
accuracy = accuracy_score(Y_test, y_pred)
accuracy

0.5452261306532663

**Answer: 0.5452261306532663**

***
### Question 4 (3 points)###
Using a NN, report the accuracy on the test set of a model that trained only on `height` and the `eyecolor` features of instances in the training set.

**Question 4.a**  
What is the accuracy on the test set using the original `height` values (no pre-processing) and `eyecolor` as a one-hot?

In [68]:
df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')

X_train_height = df_train[['height']]
X_test_height = df_test[['height']]

X_train_eyecolor = pd.get_dummies(df_train['eyecolor'], prefix='eyecolor')
X_test_eyecolor = pd.get_dummies(df_test['eyecolor'], prefix='eyecolor')


X_train_combined = pd.concat([X_train_height, X_train_eyecolor], axis=1)
X_test_combined = pd.concat([X_test_height, X_test_eyecolor], axis=1)

mlp = MLPClassifier(hidden_layer_sizes=(10,), activation='logistic', solver='lbfgs', random_state=42)
mlp.fit(X_train_combined, df_train['gender'])
y_pred = mlp.predict(X_test_combined)
accuracy_test = accuracy_score(df_test['gender'], y_pred)
accuracy_test

0.8618090452261307

**Answer: 0.8618090452261307**

**Question 4.b**  
What is the accuracy on the test set using the log of `height` values (applied to both training and testing sets) and `eyecolor` as a one-hot?

In [69]:
df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')


X_train_eyecolor = pd.get_dummies(df_train['eyecolor'], prefix='eyecolor')
X_test_eyecolor = pd.get_dummies(df_test['eyecolor'], prefix='eyecolor')
X_train_height_log = np.log1p(df_train[['height']])
X_test_height_log = np.log1p(df_test[['height']])

X_train_combined_log = pd.concat([X_train_height_log, X_train_eyecolor], axis=1)
X_test_combined_log = pd.concat([X_test_height_log, X_test_eyecolor], axis=1)

mlp_log = MLPClassifier(hidden_layer_sizes=(10,), 
                        activation='logistic', solver='lbfgs', random_state=42)
mlp_log.fit(X_train_combined_log, df_train['gender'])
y_pred_log = mlp_log.predict(X_test_combined_log)
accuracy_test_log = accuracy_score(df_test['gender'], y_pred_log)
accuracy_test_log


0.8618090452261307

**Answer: 0.8618090452261307**

**Question 4.c**  
What is the accuracy on the test set using the Z-score of `height` values and `eyecolor` as a one-hot?

Z-score is a normalization function. It is the value of a feature minus the average value for that feature (in the training set), divided by the standard deviation of that feature (in the training set). Remember that, whenever applying a function to a feature in the training set, it also has to be applied to that same feature in the test set.

In [70]:
df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')

height_mean = df_train['height'].mean()
height_std = df_train['height'].std()

X_train_height_zscore = (df_train['height'] - height_mean) / height_std
X_test_height_zscore = (df_test['height'] - height_mean) / height_std


X_train_eyecolor = pd.get_dummies(df_train['eyecolor'], 
                                  prefix='eyecolor')
X_test_eyecolor = pd.get_dummies(df_test['eyecolor'], 
                                 prefix='eyecolor')

X_train_combined_zscore = pd.concat([X_train_height_zscore, 
                                     X_train_eyecolor], axis=1)

X_test_combined_zscore = pd.concat([X_test_height_zscore, 
                                    X_test_eyecolor], axis=1)


mlp_zscore = MLPClassifier(hidden_layer_sizes=(10,), 
                           activation='logistic', solver='lbfgs', random_state=42)

mlp_zscore.fit(X_train_combined_zscore, df_train['gender'])
y_pred_zscore = mlp_zscore.predict(X_test_combined_zscore)
accuracy_test_zscore = accuracy_score(df_test['gender'], y_pred_zscore)
accuracy_test_zscore


0.8668341708542714

**Answer: 0.8668341708542714**

***
### Question 5 (1.5 points) ###
Repeat question 4 for `playgames` & `eyecolor`.

**Question 5.a**  \\
What is the accuracy on the test set using the original `playgames` values (no pre-processing) and `eyecolor` as a one-hot?

In [77]:
train_features = new_dftrain[['playgames', 'eyecolor']]
encoded_train_features = pd.get_dummies(train_features, columns=["eyecolor"])
test_features = new_dftest[['playgames', 'eyecolor']]
encoded_test_features = pd.get_dummies(test_features, columns=["eyecolor"])

X_train_features = encoded_train_features
Y_train_gender = new_dftrain['gender']

classifier_5a = MLPClassifier(hidden_layer_sizes=(10,), activation='logistic',
                    max_iter=200, solver='lbfgs', verbose=10, random_state=42)
classifier_5a.fit(X_train_features, Y_train_gender)


X_test_features = encoded_test_features
Y_test_gender = new_dftest['gender']
y_pred = classifier_5a.predict(X_test_features)
accuracy = accuracy_score(Y_test_gender, y_pred)
accuracy                           

0.678391959798995

**Answer: 0.678391959798995**

**Question 5.b**  \\
What is the accuracy on the test set using the log of `playgames` values (applied to both training and testing sets) and `eyecolor` as a one-hot?

Note: You can drop rows that have 0 in the `playgames` column, in order to avoid -inf values when applying log.

In [65]:
df_train = df_train[df_train['playgames'] > 0]
df_test = df_test[df_test['playgames'] > 0] ## --> this only selecting rows that have value greater than 0, which should be same thing?
df_test.head(1)

Unnamed: 0 gender  age    year eyecolor  height  miles  brothers  sisters  \
0        1303   male   20  second    green    73.0  210.0         0        1   

   computertime exercise  exercisehours  musiccds  playgames  watchtv  
0          10.0      Yes            5.0      50.0        1.0     15.0

In [87]:
# Training Data

train_features = new_dftrain[['playgames', 'eyecolor']]
encoded_train_features = pd.get_dummies(train_features, columns=["eyecolor"])
encoded_train_features = encoded_train_features[encoded_train_features['playgames'] > 0]
encoded_train_features['playgames'] = np.log(encoded_train_features['playgames'])


test_features = new_dftest[['playgames', 'eyecolor']]
encoded_test_features = pd.get_dummies(test_features, columns=["eyecolor"])
encoded_test_features = encoded_test_features[encoded_test_features['playgames'] > 0]
encoded_test_features['playgames'] = np.log(encoded_test_features['playgames'])


X_train_features = encoded_train_features
Y_train_gender = new_dftrain[new_dftrain['playgames'] > 0]
Y_train_gender = Y_train_gender['gender']


classifier = MLPClassifier(hidden_layer_sizes=(10,), activation='logistic',
                    max_iter=200, solver='lbfgs', verbose=10, random_state=42)
classifier.fit(X_train_features, Y_train_gender)


X_test_features = encoded_test_features
Y_test_gender = new_dftest[new_dftest['playgames'] > 0]
Y_test_gender = Y_test_gender['gender']
y_pred = classifier.predict(X_test_features)
accuracy = accuracy_score(Y_test_gender, y_pred)
accuracy


0.6300813008130082

**Answer: 0.6300813008130082**

**Question 5.c** \\
What is the accuracy on the test set using the Z-score of `playgames` values and `eyecolor` as a one-hot?

In [89]:
# YOUR CODE HERE

df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')
df_train = df_train[df_train['playgames'] > 0]
df_test = df_test[df_test['playgames'] > 0]


X_train_eyecolor = pd.get_dummies(df_train['eyecolor'], prefix='eyecolor')
X_test_eyecolor = pd.get_dummies(df_test['eyecolor'], prefix='eyecolor')
playgames_mean = df_train['playgames'].mean()
playgames_std = df_train['playgames'].std()
X_train_playgames_zscore = (df_train['playgames'] - playgames_mean) / playgames_std
X_test_playgames_zscore = (df_test['playgames'] - playgames_mean) / playgames_std
X_train_combined_playgames_zscore = pd.concat([X_train_playgames_zscore, X_train_eyecolor], axis=1)
X_test_combined_playgames_zscore = pd.concat([X_test_playgames_zscore, X_test_eyecolor], axis=1)
mlp_playgames_zscore = MLPClassifier(hidden_layer_sizes=(10,), activation='logistic',
                    max_iter=200, solver='lbfgs', verbose=10, random_state = 42)

mlp_playgames_zscore.fit(X_train_combined_playgames_zscore, df_train['gender'])
y_pred_playgames_zscore = mlp_playgames_zscore.predict(X_test_combined_playgames_zscore)
accuracy_test_playgames_zscore = accuracy_score(df_test['gender'], y_pred_playgames_zscore) 
accuracy_test_playgames_zscore

0.6757999999946341

**Answer: 0.6757999999946341**

***
### Question 6 (2 points)###
Combine the features from question 3, 4, and 5 (`year`, `eyecolor`, `exercise`, `height`, `playgames`). For numeric features use the best normalization method from questions 4 and 5.

**Question 6.a**   
What was the NN accuracy on the test set using the single 10 node hidden layer?

In [90]:
# YOUR CODE HERE

df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')
df_train = df_train[df_train['playgames'] > 0]
df_test = df_test[df_test['playgames'] > 0]


height_mean = df_train['height'].mean()
height_std = df_train['height'].std()
X_train_height_zscore = (df_train['height'] - height_mean) / height_std
X_test_height_zscore = (df_test['height'] - height_mean) / height_std


playgames_mean = df_train['playgames'].mean()
playgames_std = df_train['playgames'].std()
X_train_playgames_zscore = (df_train['playgames'] - playgames_mean) / playgames_std
X_test_playgames_zscore = (df_test['playgames'] - playgames_mean) / playgames_std
categorical_features = ['year', 'eyecolor', 'exercise']
X_train_categorical = pd.get_dummies(df_train[categorical_features], columns=categorical_features, prefix=categorical_features)
X_test_categorical = pd.get_dummies(df_test[categorical_features], columns=categorical_features, prefix=categorical_features)
X_train_combined = pd.concat([X_train_height_zscore, X_train_playgames_zscore, X_train_categorical], axis=1)
X_test_combined = pd.concat([X_test_height_zscore, X_test_playgames_zscore, X_test_categorical], axis=1)

mlp_combined = MLPClassifier(hidden_layer_sizes=(10,), activation='logistic', solver='lbfgs', random_state=42)
mlp_combined.fit(X_train_combined, df_train['gender'])

y_pred_combined = mlp_combined.predict(X_test_combined)
accuracy_test_combined = accuracy_score(df_test['gender'], y_pred_combined)
accuracy_test_combined

0.8189999999927642

**Answer: 0.8189999999927642**

***
### Question 7- Bonus (1 point)###
Can you improve your test set prediction accuracy by 3% or more? See how close to that milestone of improvement you can get by modifying the hyperparameters of  Neural Networks (the number of hidden layers, number of hidden nodes in each layer, the learning rate, the type of activation function etc.).

A great guide to tuning parameters is explained in this guide: http://www.csie.ntu.edu.tw/~cjlin/papers/guide/guide.pdf. While the guide is specific to SVM and in particular the C and gamma parameters of the RBF kernel, the method applies more generally to any ML technique with tuning parameters.

Please give your new prediction accuracy on the test set, and write a paragraph in a text cell below with an explanation of your approach and evaluation metrics.


**Answer: YOUR ANSWER HERE**

**Explanation: YOUR ANSWER HERE**